storage/redis/peer_store.go

```diff
@@ -0,0 +1,787 @@
+	IPver := p.IP.AddressFamily.String()
+	log.Debug("storage: DeleteLeecher", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), IPver, p.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	encoded_ih_L := IPver + "_L_" + ih.String()
+
+	DelNum, err := conn.Do("HDEL", encoded_ih_L, pk)
+	if err != nil {
+		return err
+	}
+	if DelNum.(int64) == 0 {
+		return storage.ErrResourceDoesNotExist
+	}
+
+	return nil
+}
+
+func (ps *peerStore) GraduateLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	IPver := p.IP.AddressFamily.String()
+	log.Debug("storage: GraduateLeecher", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), IPver, p.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih
+	encoded_ih_S := IPver + "_S_" + encoded_ih
+
+	_, err := conn.Do("HDEL", encoded_ih_L, pk)
+	if err != nil {
+		return err
+	}
+
+	// Update the peer in the swarm.
+	ct := ps.getClock()
+	_, err = conn.Do("HSET", encoded_ih_S, pk, ct)
+
+	if err != nil {
+		return err
+	}
+
+	_, err = conn.Do("HSET", IPver, encoded_ih_S, ct)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *peerStore) AnnouncePeers(ih bittorrent.InfoHash, seeder bool, numWant int, announcer bittorrent.Peer) (peers []bittorrent.Peer, err error) {
+	IPver := announcer.IP.AddressFamily.String()
+	log.Debug("storage: AnnouncePeers", log.Fields{
+		"InfoHash": ih.String(),
+		"seeder":   seeder,
+		"numWant":  numWant,
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", announcer.ID.String(), announcer.IP.String(), IPver, announcer.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih // key
+	encoded_ih_S := IPver + "_S_" + encoded_ih // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechers, err := conn.Do("HKEYS", encoded_ih_L)
+	if err != nil {
+		return nil, err
+	}
+	con_leechers := leechers.([]interface{})
+
+	seeders, err := conn.Do("HKEYS", encoded_ih_S)
+	if err != nil {
+		return nil, err
+	}
+	con_seeders := seeders.([]interface{})
+
+	if len(con_leechers) == 0 && len(con_seeders) == 0 {
+		return nil, storage.ErrResourceDoesNotExist
+	}
+
+	if seeder {
+		// Append leechers as possible.
+		for _, pk := range con_leechers {
+			if numWant == 0 {
+				break
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
```

Comment:
Is it guaranteed that these type assertions never blow up or panic?

Reply:
I cannot guarantee, but It should work fine as far as I have learned.